pandas/_libs/tslibs/fields.pyx

```diff
@@ -43,6 +48,36 @@ def get_time_micros(ndarray[int64_t] dtindex):
     return micros
 
 
+@cython.wraparound(False)
+@cython.boundscheck(False)
+def get_datetime_isoformats(ndarray[int64_t] dtindex) -> ndarray:
+    """
+    Return isoformats for an array of datetimelike objects.
+
+    Parameters
+    ----------
+    dtindex : DatetimeArray
+
+    Returns
+    -------
+    Array of ISO formats
+    """
+    cdef:
+        Py_ssize_t i, count = len(dtindex)
+        int64_t val, convert_status
+        npy_datetimestruct dts
+        char buf[34]  # ns precision with UTC offset max length
+        
+    out = np.empty(count, dtype=object)
+
+    for i in range(count):
+        dt64_to_dtstruct(dtindex[i], &dts);
+        # TODO: handle bad return
+        convert_status = make_iso_8601_datetime(&dts, buf, 34, NPY_FR_ns)
+        out[i] = buf.decode("UTF-8")
```

Comment:
is returning this as bytes essential?

side-thought: could tweak perf by making `out = np.empty(count, dtype="S34")`?

Reply:
`make_iso_8601_datetime` operates on C strings so this converts the C string to a unicode object. I find this a little surprising that this is how it works but was pulled directly from Cython docs

http://docs.cython.org/en/latest/src/tutorial/strings.html#decoding-bytes-to-text